### PR TITLE
Instrument all allocation paths

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

Resolves: #1617
Resolves: #1635 

## Description

Update memory allocation instrumentation to catch additional paths.

## Testing

CI/CD

## Documentation

_Is there any documentation impact for this change?_

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
Converting to draft until I find why this breaks the fuzzing tests.

**Comment by Alan-Jowett** [comment]
std::vector's default constructor is noexcept. The problem is that the Microsoft STL implementation can throw bad_alloc from the default constructor in debug mode. Net result is that:
```
std::vector<foo> bar;
```
Will terminate under low memory in debug mode, which makes our current implementation of new/delete instrumentation super noisy.

**Comment by Alan-Jowett** [comment]
Need to set iterator debug level to zero to disable this:
https://learn.microsoft.com/en-us/cpp/standard-library/debug-iterator-support?view=msvc-170

File: libs/platform/unit/platform_unit_test.cpp [hunk 1]
```diff
@@ -631,15 +631,16 @@ TEST_CASE("serialize_map_test", "[platform]")
...
+            map_count, internal_map_info_array, buffer, buffer_length, &serialized_length, &required_length) ==
+        EBPF_INSUFFICIENT_BUFFER);
 
-    buffer = static_cast<uint8_t*>(calloc(required_length, 1));
-    REQUIRE(buffer != nullptr);
+    buffer = static_cast<uint8_t*>(ebpf_allocate(required_length));
     if (!buffer) {
         return;
     }
+    REQUIRE(buffer != nullptr);
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:643]: not needed since if it were nullptr then execution would return at 641

**Review by dthaler — DISMISSED** [review]

**Review by dthaler — DISMISSED** [review]

**Review by dthaler — DISMISSED** [review]

File: libs/api/ebpf_api.cpp [hunk 2]
```diff
@@ -1061,11 +1061,8 @@ _clean_up_ebpf_link(_Frees_ptr_opt_ ebpf_link_t* link) noexcept
     if (link->fd != ebpf_fd_invalid) {
         Platform::_close(link->fd);
     }
-
     free(link->pin_path);
-
     free(link);
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:1065]: Should we change this and the one on line 1064 to `ebpf_free`?

File: libs/api/ebpf_api.cpp [hunk 3]
```diff
@@ -1061,11 +1061,8 @@ _clean_up_ebpf_link(_Frees_ptr_opt_ ebpf_link_t* link) noexcept
...
-#pragma warning(pop)
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:1068]: Is removing this intentional?

File: libs/platform/unit/platform_unit_test.cpp [hunk 4]
```diff
@@ -634,13 +634,15 @@ TEST_CASE("serialize_map_test", "[platform]")
...
+            map_count, internal_map_info_array, buffer, buffer_length, &serialized_length, &required_length) ==
+        EBPF_INSUFFICIENT_BUFFER);
 
-    buffer = static_cast<uint8_t*>(calloc(required_length, 1));
-    REQUIRE(buffer != nullptr);
-    if (!buffer) {
+    buffer = static_cast<uint8_t*>(ebpf_allocate(required_length));
+    // Required to deal with code analysis warning about buffer not being checked for null.
+    if (buffer == nullptr) {
+        REQUIRE(false);
```
Comment (saxena-anurag) [inline libs/platform/unit/platform_unit_test.cpp:645]: nit: can we change it back to `REQUIRE(buffer != nullptr)` ?

File: libs/platform/unit/platform_unit_test.cpp [hunk 5]
```diff
@@ -699,8 +701,12 @@ TEST_CASE("serialize_program_info_test", "[platform]")
     // Serialize.
     REQUIRE(ebpf_serialize_program_info(&in_program_info, buffer, buffer_length, &serialized_length, &required_length));
 
-    buffer = static_cast<uint8_t*>(calloc(required_length, 1));
-    _Analysis_assume_(buffer != nullptr);
+    buffer = static_cast<uint8_t*>(ebpf_allocate(required_length));
+    // Work around code analysis warning about buffer not being checked for null.
+    if (buffer == nullptr) {
+        REQUIRE(false);
+        return;
```
Comment (saxena-anurag) [inline libs/platform/unit/platform_unit_test.cpp:708]: same as earlier

File: libs/platform/unit/platform_unit_test.cpp (hunk 4 above)
Comment (Alan-Jowett) [reply libs/platform/unit/platform_unit_test.cpp:645]: Unfortunately not. The code analysis tool doesn't see REQUIRE(buffer != nullptr) as observing the buffer handle.

File: libs/platform/unit/platform_unit_test.cpp (hunk 5 above)
Comment (Alan-Jowett) [reply libs/platform/unit/platform_unit_test.cpp:708]: Same as earlier.

**Review by dthaler — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**